integrations: Add support for "Test plugin" in Sentry integration.

sbansal1999 opened the pull request:
If a user tries to create a webhook using the Webhooks plugin in Sentry and uses the "Test plugin" to test the webhook, the server would send a 500 error, even though the integration worked perfectly. This led users to believe that the integration was not working. 

This PR fixes that behaviour and makes sure the sample event is handled properly.

Fixes: #26173 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
@alexmv can you review this one, since you know more about Sentry than I do?

alexmv commented on zerver/webhooks/sentry/view.py:
```diff
@@ -81,6 +81,20 @@
 
 
+def is_sample_event(event: Dict[str, Any]) -> bool:
+    # These are just some workarounds; there is no documented way to check
+    # if an event is a sample event or not.
+    tags = event.get("tags", [])
+    if ["sample_event", "yes"] in tags:
+        return True
+    title = event.get("title", "")
+    if title == "This is an example Python exception":
+        return True
```
I'd love a mention in here of why we have both of these checks, since they seem potentially redundant unless one knows the history here.

sbansal1999 replied:
You are right the checks are redundant. I would remove the `tags` check and just keep the `title` check for sample events.

> Does the Sentry UI show the response body from the server when you push the "test" button? 

When the integration is set through Alerts there is no response body shown to the user but when it is set through the Webhook plugin a response body is shown only when an error is returned. So I don't think we can hint the user as such. I think since it's clearly mentioned in the [documentation](https://zulip.com/integrations/doc/sentry) that setting the integration through the Webhook plugin is not the best idea hence, there is no need for this.

>  when they're setting it up via alerts, which may not provide the full datastream

I think you meant to say via the webhook plugin because if someone sets an integration through that that's when the integration doesn't provide all the events.

alexmv replied:
I actually wasn't suggesting that you remove the other check -- I said they "seem potentially redundant unless one knows the history here."  I feel that, if we could only use it, the tags solution is better because it's less likely to have false-positives.  I can see users typing `raise("This is an example Python exception")` in their code when trying it out, and I'd rather that we not _think_ it's s sample-from-Sentry.  Right now, the only effect is that we'd accept a Raven SDK submission, but code does drift and if we, say, started doing other things based on `is_sample_event`, false-positives would get more dangerous.

Since we _have_ to use the title, I'm fine with dropping the tag support, but I'd prefer to adjust the "This is just a workaround" comment to make clearer that this is a _heuristic_ which may false-positive, so we should not rely on it for important behaviour decisions.

> When the integration is set through Alerts there is no response body shown to the user but when it is set through the Webhook plugin a response body is shown only when an error is returned. So I don't think we can hint the user as such.

Bummer.

> I think since it's clearly mentioned in the [documentation](https://zulip.com/integrations/doc/sentry) that setting the integration through the Webhook plugin is not the best idea hence, there is no need for this.

It's documented, but _people will not always read the documentation_ -- which is what made us need this PR in the first place. :)  So we should always think about if it's simple to have the user interface (here, through the Sentry UI) give good feedback.  So while there's no "need" for this, in that it's not a blocker, it certainly would have improved the product experience if we could.

> I think you meant to say via the Webhook plugin because if someone sets an integration through that that's when the integration doesn't provide all the events.

Ah, yeah, I said that backwards.  Sorry for any confusion!

sbansal1999 replied:
> Since we have to use the title, I'm fine with dropping the tag support, but I'd prefer to adjust the "This is just a workaround" comment to make clearer that this is a heuristic which may false-positive, so we should not rely on it for important behaviour decisions.

Yes, I think that makes perfect sense.

>So while there's no "need" for this, in that it's not a blocker, it certainly would have improved the product experience if we could.

You are correct that would have certainly improved the overall experience. Since there is no way right now to do that, I feel we can leave this for now.

alexmv commented on zerver/webhooks/sentry/view.py:
```diff
@@ -81,6 +81,20 @@
+        return True
+    # This happens if someday Sentry decides to change the title of the
+    # sample event or remove the [sample_event, yes] tag.
+    return False  # nocoverage
```
Doesn't this nocoverage line means that we don't test with any of the old-style Raven data?  Can we add a fixture and test for a non-sample-event one of those?

sbansal1999 replied:
Agreed. We should add an old-style test to the integration.

sbansal1999 commented:
**Updates:**

- Added the test for Raven SDK for python.
- Removed a redundant sample event check as suggested in [this](https://github.com/zulip/zulip/pull/26231#discussion_r1261590784) comment.

sbansal1999 commented:
**Updates:**
- Updated the comment in the `is_sample_event` function to make it clear that the function should not be used to make important behavior decisions.

sbansal1999 commented:
@alexmv Sorry for the delay on this one. I have updated the PR; let me know if there is anything more required in this one.

alexmv reviewed (commented):
Thanks, @sbansal1999 -- this looks good to me now!

timabbott commented:
Merged, and marked as a backport candidate, thanks @sbansal1999 and @alexmv!

alexmv commented:
Backported to 7.x as d3048b8791d45b324e5ca9ad40e949fd04cfb7d0 through 35b0b2fb6a44ea86870a7e24a3a89da9474489bb, with bonus backports of #25861, #26224, #26230, and #24983, since this depends on the first and last of those, and the middle two make sense to include.
